src/servers/src/http/event.rs

```diff
@@ -352,6 +362,163 @@ pub async fn pipeline_dryrun(
+    TypedHeader(content_type): TypedHeader<ContentType>,
+    LogTableName(table_name): LogTableName,
+    bytes: Bytes,
+) -> Result<HttpResponse> {
+    ctx.set_channel(Channel::Loki);
+    let ctx = Arc::new(ctx);
+    let db = ctx.get_db_string();
+    let db_str = db.as_str();
+    let table_name = table_name.unwrap_or("loki_logs".to_string());
+    let exec_timer = Instant::now();
+
+    // decompress req
+    ensure!(
+        &content_type.to_string() == "application/x-protobuf",
+        UnsupportedContentTypeSnafu { content_type }
+    );
+    let decompressed = prom_store::snappy_decompress(&bytes).unwrap();
+    let req = loki_api::logproto::PushRequest::decode(&decompressed[..])
+        .context(DecodeOtlpRequestSnafu)?;
+
+    // init schemas
+    let mut schemas: Vec<ColumnSchema> = vec![
+        ColumnSchema {
+            column_name: "timestamp".to_string(),
+            datatype: ColumnDataType::TimestampNanosecond.into(),
+            semantic_type: SemanticType::Timestamp.into(),
+            datatype_extension: None,
+            options: None,
+        },
+        ColumnSchema {
+            column_name: "line".to_string(),
+            datatype: ColumnDataType::String.into(),
+            semantic_type: SemanticType::Field.into(),
+            datatype_extension: None,
+            options: None,
+        },
+    ];
+
+    let mut global_label_key_index: HashMap<String, i32> = HashMap::new();
+    global_label_key_index.insert("timestamp".to_string(), 0);
+    global_label_key_index.insert("line".to_string(), 1);
+
+    let mut rows = vec![];
+
+    for stream in req.streams {
+        // parse labels for each row
+        // encoding: https://github.com/grafana/alloy/blob/be34410b9e841cc0c37c153f9550d9086a304bca/internal/component/common/loki/client/batch.go#L114-L145
+        // use very dirty hack to parse labels
+        let labels = stream.labels.replace("=", ":");
+        // use btreemap to keep order
+        let labels: BTreeMap<String, String> = json5::from_str(&labels).context(ParseJson5Snafu)?;
+
+        // process entries
+        for entry in stream.entries {
+            let ts = if let Some(ts) = entry.timestamp {
+                ts
+            } else {
+                continue;
+            };
+            let line = entry.line;
+
+            // create and init row
+            let mut row = Vec::with_capacity(schemas.capacity());
+            for _ in 0..row.capacity() {
+                row.push(GreptimeValue { value_data: None });
+            }
+            // insert ts and line
+            row[0] = GreptimeValue {
+                value_data: Some(ValueData::TimestampNanosecondValue(
+                    ts.seconds * 1000000000 + ts.nanos as i64,
+                )),
+            };
+            row[1] = GreptimeValue {
+                value_data: Some(ValueData::StringValue(line)),
+            };
+            // insert labels
+            for (k, v) in labels.iter() {
```

Comment:
One Loki push request contains multiple log entries. The labels are request-level attributes, so it would be appended to each row of log entry. Hence the `iter()` instead of `into_iter()`.